course/include/isometry.h

```diff
@@ -0,0 +1,113 @@
+  bool operator == (const Vector3 & obj) const;
+  bool operator != (const std::initializer_list<double> & obj) const {return !(*this == obj);};
+  bool operator != (const Vector3 & obj) const {return !(*this == obj);};
+  Vector3 operator *= (const double & obj) const;
+  Vector3 operator /= (const double & obj) const;
+  Vector3 operator * (const double & obj) const {return *this *= obj;};
+  Vector3 operator / (const double & obj) const {return *this /= obj;};
+
+  Vector3 cross(const Vector3 & obj) const;
+  double dot(const Vector3 & obj) const;
+  double norm() const;
+
+  friend std::ostream& operator << (std::ostream & os, const Vector3 & v) {
+            os << std::string("(x: ") << v.x() << ", y: " << v.y() << ", z: " << v.z() << ")";
+            return os;
+        };
+ private:
+   double x_,y_,z_;
+};
+
+Vector3 operator * (const double & obj1, const Vector3 & obj2);
+
+class Matrix3 {
+  public:
+    static const Matrix3 kIdentity;
+    static const Matrix3 kZero;
+    static const Matrix3 kOnes;
+    Matrix3(std::initializer_list<double> row1, std::initializer_list<double> row2,
+       std::initializer_list<double> row3): row1_(row1), row2_(row2), row3_(row3) {};
+    Matrix3(double x1, double y1, double z1, double x2, double y2, double z2,
+       double x3, double y3, double z3): row1_(x1, y1, z1), row2_(x2, y2, z2), row3_(x3, y3, z3) {};
+    Matrix3(const Vector3 vec1, const Vector3 vec2, const Vector3 vec3): row1_(vec1), row2_(vec2), row3_(vec3) {};
+    Matrix3(const Matrix3 & obj): row1_(obj[0]), row2_(obj[1]), row3_(obj[2]) {};
+    Matrix3(): row1_(0., 0., 0.), row2_(0., 0., 0.), row3_(0., 0., 0.) {};
+    Vector3 row(const int index) const;
+    Vector3 col(const int index) const;
+    bool operator == (const Matrix3 & obj) const;
+    bool operator != (const Matrix3 & obj) const {return !(*this == obj);};
+    bool operator == (const std::initializer_list<double> & obj) const;
+    bool operator != (const std::initializer_list<double> & obj) const {return !(*this == obj);};
+    Matrix3 operator += (const Matrix3 & obj) const;
+    Matrix3 operator -= (const Matrix3 & obj) const;
+    Matrix3 operator *= (const Matrix3 & obj) const;
+    Matrix3 operator /= (const Matrix3 & obj) const;
+    Matrix3 operator + (const Matrix3 & obj) const {return *this += obj;};
+    Matrix3 operator - (const Matrix3 & obj) const {return *this -= obj;};
+    Matrix3 operator * (const Matrix3 & obj) const {return *this *= obj;};
+    Matrix3 operator / (const Matrix3 & obj) const {return *this /= obj;};
+    Matrix3 operator *= (const double & obj) const;
+    Matrix3 operator /= (const double & obj) const;
+    Matrix3 operator * (const double & obj) const {return *this *= obj;};
+    Matrix3 operator / (const double & obj) const {return *this /= obj;};
+    Vector3 operator [] (const int index) const;
+    Vector3 & operator [] (const int index);
+    double det() const;
+    Vector3 dot(const Vector3 & obj) const;
```

Comment:
There's no `dot` product between matrix and vector. That's just the normal matrix product. You need to overload `Vector3 operator * (const Vector3 & obj) const`